[PyTorch Pin Memory Allocator] Optimize the free list implementation and add lock sharding

banitag1 opened the pull request:
Summary: This diff addresses the lock contention issue in free list implementation of CachingHost/Pinned allocator. We add a different data structure for free list and also add lock sharding based on allocation size.

Differential Revision: D61623367

guangyey commented on aten/src/ATen/core/CachingHostAllocator.h:
```diff
@@ -246,13 +249,12 @@ struct CachingHostAllocatorImpl {
-    auto it = free_list_.lower_bound(&key);
-    if (it != free_list_.end()) {
-      B* block = *it;
+    auto index = size_index(size);
```
If I'm not mistaken, this indeed changes the logic of finding the suitable free block. Previously, the entire free list was searched for an appropriate lock; for example, if 64 bytes were requested but only 128 bytes were available in the free list, the 128 bytes block would be reused. However, in the new implementation, a new 64 bytes allocation would be performed instead of reusing the 128 bytes block.
The new implementation seems more reasonable since the host caching allocator doesn't split the excess blocks to improve storage utilization.
If you could share the data on the performance or memory utilization improvements, that would be great. Others look good to me.

banitag1 replied:
Based on some initial testing, we didn't see much difference in the memory, it was mostly in the noise margin.

EikanWang commented:
IIUC, this PR buckets the free memory blocks. Compared to the original flat implementation, there are two layers now for the free memory block list. The top layer is bucketed by the leading zero of a given size, while each bucket holds a range of free memory blocks. Suppose a particular size with `n` leading zeros, all the free memory blocks will be part of the free memory block bucket if the memory block size is between `(2**(n-1), 2**n]`.

If `n` is large, the granularity is not the same as the original implementation. Because the original implementation always picks up the block whose size is close to the memory size to be allocated - `free_list_.lower_bound(&key);.`
However, the current implementation always picks up the last block - `B* block = free_list_[index].list_.back();`

For example, there are 2 free memory blocks, and their sizes are as follows. The two blocks will be bucketed to the same index.

- block 0: 2048
- block 1: 4096

Suppose a memory allocation order:

- Step1: Alloc 2048
- Step2: Alloc 4096

The current implementation has to allocate another 4096 memory because Step 1 has consumed block 1, and then Step 2 cannot be met.

Is my understanding correct? @banitag1

albanD commented on aten/src/ATen/core/CachingHostAllocator.h:
```diff
@@ -337,12 +340,11 @@ struct CachingHostAllocatorImpl {
+  // size. This allows us to quickly find a free block of the right size.
+  // We use deque to store per size free list and guard the list with its own
+  // mutex.
+  alignas(64) std::vector<FreeBlockList<B>> free_list_ = std::vector<FreeBlockList<B>>(MAX_SIZE_INDEX);
```
Why the alignment constraint on this vector?

albanD commented on aten/src/ATen/core/CachingHostAllocator.h:
```diff
@@ -30,20 +30,17 @@ struct HostBlock {
-    }
-    return (uintptr_t)a->ptr_ < (uintptr_t)b->ptr_;
-  }
+struct alignas(64) FreeBlockList {
```
Why the extra alignment constraint here?

banitag1 replied:
> Why the extra alignment constraint here?

This is to avoid cache line false sharing as we have a mutex in the free list struct. So, we want to give exclusive cache line to the mutex to avoid lots of cache line ping-ponging across CPU cores.

albanD replied:
Do we have a better value than the magic 64bytes being used here given the wide range of CPU pytorch runs on?

EikanWang commented:
> IIUC, this PR buckets the free memory blocks. Compared to the original flat implementation, there are two layers now for the free memory block list. The top layer is bucketed by the leading zero of a given size, while each bucket holds a range of free memory blocks. Suppose a particular size with `n` leading zeros, all the free memory blocks will be part of the free memory block bucket if the memory block size is between `(2**(n-1), 2**n]`.
> 
> If `n` is large, the granularity is not the same as the original implementation. Because the original implementation always picks up the block whose size is close to the memory size to be allocated - `free_list_.lower_bound(&key);.` However, the current implementation always picks up the last block - `B* block = free_list_[index].list_.back();`
> 
> For example, there are 2 free memory blocks, and their sizes are as follows. The two blocks will be bucketed to the same index.
> 
> * block 0: 2048
> * block 1: 4096
> 
> Suppose a memory allocation order:
> 
> * Step1: Alloc 2048
> * Step2: Alloc 4096
> 
> The current implementation has to allocate another 4096 memory because Step 1 has consumed block 1, and then Step 2 cannot be met.
> 
> Is my understanding correct? @banitag1

I see. Since each bucket always holds the free memory block of the same size. The example does not exist.

banitag1 commented:
> IIUC, this PR buckets the free memory blocks. Compared to the original flat implementation, there are two layers now for the free memory block list. The top layer is bucketed by the leading zero of a given size, while each bucket holds a range of free memory blocks. Suppose a particular size with `n` leading zeros, all the free memory blocks will be part of the free memory block bucket if the memory block size is between `(2**(n-1), 2**n]`.
> 
> If `n` is large, the granularity is not the same as the original implementation. Because the original implementation always picks up the block whose size is close to the memory size to be allocated - `free_list_.lower_bound(&key);.` However, the current implementation always picks up the last block - `B* block = free_list_[index].list_.back();`
> 
> For example, there are 2 free memory blocks, and their sizes are as follows. The two blocks will be bucketed to the same index.
> 
> * block 0: 2048
> * block 1: 4096
> 
> Suppose a memory allocation order:
> 
> * Step1: Alloc 2048
> * Step2: Alloc 4096
> 
> The current implementation has to allocate another 4096 memory because Step 1 has consumed block 1, and then Step 2 cannot be met.
> 
> Is my understanding correct? @banitag1

Currently all the allocations are rounded to power of 2, which makes it easier to also manage the free list.
size_t roundSize = c10::llvm::PowerOf2Ceil(size);

In the current implementation, step 1 allocation of 2048 bytes will never use the  allocation from 4096 size free list. if there is none available in 2048 size free list, it will allocate a new one.

banitag1 committed:
[PyTorch Pin Memory Allocator] Optimize the free list implementation and add lock sharding (#134154)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134154

This diff addresses the lock contention issue in free list implementation of CachingHost/Pinned allocator. We add a different data structure for free list and also add lock sharding based on allocation size.

Reviewed By: zyan0

Differential Revision: D61623367

banitag1 commented:
@pytorchbot label "topic: not user facing"

banitag1 commented:
@pytorchbot merge

banitag1 commented:
@pytorchbot label "topic: not user facing"
